File: backend/src/domain/content/domain.rs
#![allow(dead_code)]
use std::sync::Arc;

use crate::{
    datasource::{dbdao::dao::DBDao, scylladao::dao::ScyllaDao, vectordao::dao::VectorDao},
    domain::content::schema::Content,
    infra::registry::Registry,
};

#[derive(Clone)]
pub struct ContentDomain {
    pub db_dao: Arc<DBDao>,
    pub scylla_dao: Arc<ScyllaDao>,
    pub vector_dao: Arc<VectorDao>,
}

impl ContentDomain {
    pub fn new(r: &Registry) -> Self {
        Self {
            db_dao: r.db_dao.clone(),
            scylla_dao: r.scylla_dao.clone(),
            vector_dao: r.vector_dao.clone(),
        }
    }

    pub async fn get_content(&self, id: i64) -> Option<Content> {
        self.db_dao
            .get_content(id)
            .await
            .ok()
            .map(|record| Content {
                id: record.id,
                title: record.title,
                content: record.content,
                content_type: record.content_type,
                status: record.status,
                tags: record
                    .tags
                    .map(|s| s.split(',').map(|t| t.trim().to_string()).collect()),
                created_at: record.created_at,
                updated_at: record.updated_at,
                published_at: record.published_at,
            })
    }

    pub async fn search_contents(
        &self,
        query: Option<String>,
        content_type: Option<String>,
        status: Option<String>,
        tag: Option<String>,
        page: i64,
        limit: i64,
    ) -> (Vec<Content>, i64) {
        let query = query.map(|q| format!("%{}%", q));
        match self
            .db_dao
            .search_contents(query.as_deref(), content_type, status, tag, page, limit)
            .await
        {
            Ok(records) => {
                let contents: Vec<Content> = records
                    .into_iter()
                    .map(|record| Content {
                        id: record.id,
                        title: record.title,
                        content: record.content,
                        content_type: record.content_type,
                        status: record.status,
                        tags: record
                            .tags
                            .map(|s| s.split(',').map(|t| t.trim().to_string()).collect()),
                        created_at: record.created_at,
                        updated_at: record.updated_at,
                        published_at: record.published_at,
                    })
                    .collect();
                let total = contents.len() as i64;
                (contents, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn add_content(
        &self,
        title: String,
        content: String,
        content_type: Option<String>,
        tags: Option<Vec<String>>,
    ) -> Result<i64, String> {
        let tags = tags.map(|v| v.join(","));
        self.db_dao
            .add_content(title.as_str(), content.as_str(), content_type, tags)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_content(
        &self,
        id: i64,
        title: Option<String>,
        content: Option<String>,
        content_type: Option<String>,
        tags: Option<Vec<String>>,
        _status: Option<String>,
    ) -> Result<i64, String> {
        let tags = tags.map(|v| v.join(","));
        self.db_dao
            .update_content(id, title, content, content_type, tags)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn delete_content(&self, id: i64) -> Result<i64, String> {
        self.db_dao
            .delete_content(id)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn create_content_version(
        &self,
        content_id: i64,
        content: String,
        created_by: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .create_content_version(content_id, content.as_str(), created_by)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn publish_content(
        &self,
        id: i64,
        published_at: Option<String>,
        published_by: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .publish_content(id, published_at, published_by)
            .await
            .map_err(|e| e.to_string())
    }
}

File: backend/domain/content/domain.go
package content

import (
	"backend/datasource/dbdao"
	"context"
	"errors"
)

type ContentDomain struct {
	DB *dbdao.DB
}

func NewContentDomain(db *dbdao.DB) *ContentDomain {
	return &ContentDomain{DB: db}
}

// GetContent 获取内容详情
func (d *ContentDomain) GetContent(ctx context.Context, id int64) (*Content, error) {
	content, err := d.DB.GetContent(ctx, id)
	if err != nil {
		return nil, err
	}

	return &Content{
		ID:          content.ID,
		Title:       content.Title,
		Body:        content.Body,
		ContentType: content.ContentType,
		Status:      content.Status,
		Tags:        content.Tags,
		CreatedAt:   &content.CreatedAt,
		UpdatedAt:   &content.UpdatedAt,
		PublishedAt: content.PublishedAt,
	}, nil
}

// SearchContents 搜索内容
func (d *ContentDomain) SearchContents(
	ctx context.Context,
	query, contentType, status, tag string,
	page, limit int64,
) ([]*Content, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}

	contents, total, err := d.DB.SearchContents(ctx, query, contentType, status, tag, page, limit)
	if err != nil {
		return nil, 0, err
	}

	result := make([]*Content, len(contents))
	for i, c := range contents {
		result[i] = &Content{
			ID:          c.ID,
			Title:       c.Title,
			Body:        c.Body,
			ContentType: c.ContentType,
			Status:      c.Status,
			Tags:        c.Tags,
			CreatedAt:   &c.CreatedAt,
			UpdatedAt:   &c.UpdatedAt,
			PublishedAt: c.PublishedAt,
		}
	}

	return result, total, nil
}

// AddContent 添加内容
func (d *ContentDomain) AddContent(
	ctx context.Context,
	title, body, contentType, status string,
	tags []string,
) (*Content, error) {
	if title == "" || body == "" || contentType == "" {
		return nil, errors.New("missing required fields")
	}

	content := &dbdao.Content{
		Title:       title,
		Body:        body,
		ContentType: contentType,
		Status:      status,
		Tags:        tags,
	}

	created, err := d.DB.CreateContent(ctx, content)
	if err != nil {
		return nil, err
	}

	return &Content{
		ID:          created.ID,
		Title:       created.Title,
		Body:        created.Body,
		ContentType: created.ContentType,
		Status:      created.Status,
		Tags:        created.Tags,
		CreatedAt:   &created.CreatedAt,
		UpdatedAt:   &created.UpdatedAt,
		PublishedAt: created.PublishedAt,
	}, nil
}

// UpdateContent 更新内容
func (d *ContentDomain) UpdateContent(
	ctx context.Context,
	id int64,
	title, body, contentType, status *string,
	tags []string,
) (*Content, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}

	content := &dbdao.Content{
		ID:   id,
		Tags: tags,
	}

	if title != nil {
		content.Title = *title
	}
	if body != nil {
		content.Body = *body
	}
	if contentType != nil {
		content.ContentType = *contentType
	}
	if status != nil {
		content.Status = *status
	}

	updated, err := d.DB.UpdateContent(ctx, content)
	if err != nil {
		return nil, err
	}

	return &Content{
		ID:          updated.ID,
		Title:       updated.Title,
		Body:        updated.Body,
		ContentType: updated.ContentType,
		Status:      updated.Status,
		Tags:        updated.Tags,
		CreatedAt:   &updated.CreatedAt,
		UpdatedAt:   &updated.UpdatedAt,
		PublishedAt: updated.PublishedAt,
	}, nil
}

// DeleteContent 删除内容
func (d *ContentDomain) DeleteContent(ctx context.Context, id int64) error {
	if id <= 0 {
		return errors.New("invalid id")
	}
	return d.DB.DeleteContent(ctx, id)
}
